### PR TITLE
Feature Branch With all the created methods

### PR DESCRIPTION
**Pull request by o-rashad**
Methods Created with the different Ruby tools :

1. my_each
1. my_each_with_index
1. my_select
1. my_all?
1. my_any?
1. my_none?
1. my_conut
1. my_map
1. my_inject

1. my_mulitple_els

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,121 @@
+module Enumerable
+  # my_each()
+  def my_each()
+    each do |item|
+      yield item if block_given?
+    end
+  end
```
Comment (mahmoud717) [inline main.rb]: - [x] my_each returns an Enumerator if no block is given

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  # ending my_each()
+
+  # my_each_with_index()
+  def my_each_with_index()
+    index = 0
+    each do |item|
+      yield(item, index) if block_given?
+      index += 1
+    end
+  end
```
Comment (mahmoud717) [inline main.rb]: - [x] my_each_with_index returns an enumerator if no block is given

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  # my_each_with_index()
+
+  # my_select
+  def my_select()
+    new_arr = []
+    each do |item|
+      new_arr << item if yield(item)
+    end
+    new_arr
+  end
```
Comment (mahmoud717) [inline main.rb]: - [x] my_select returns an enumerator if no block is given

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+  # my_all?
+  def my_all?()
+    flag = true
+    each do |item|
+      feedback = yield(item)
+      flag = false if (feedback == false) || feedback.nil?
+    end
+    flag
+  end
```
Comment (mahmoud717) [inline main.rb]: - [x] my_all? when no block or argument is given returns true when none of the collection members are false or nil

- [x] my_all? when no block or argument is given returns false when one of the collection members are false or nil

- [x] my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer

- [x] enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric

- [x] enumerables #my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class

- [x] enumerables #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex

- [x] my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex

- [x] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

- [x] my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  # my_all?
+
+  # my_any
+  def my_any?()
+    flag = false
+    each do |item|
+      flag = true if yield(item)
+    end
+    flag
+  end
```
Comment (mahmoud717) [inline main.rb]: - [x] my_any? when no block or argument is given returns false if at least one of the collection is not true

- [x] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric

- [x] my_any? when no block or argument is given returns true if at least one of the collection is not false or nil

- [x] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer

- [x] my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex

- [x] my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex

- [x] my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

- [x] my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  # my_any
+
+  # my_none?
+  def my_none?()
+    is_none = true
+    each do |item|
+      is_none = false if yield(item)
+    end
+    is_none
+  end
```
Comment (mahmoud717) [inline main.rb]: - [x] my_none? when no block or argument is given returns true only if none of the collection members is true

- [x] my_none? when no block or argument is given returns false only if one of the collection members is true

- [x] my_none? when a class is passed as an argument returns true if none of the collection is a member of such class

- [x] my_none? when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric

- [x] my_none? when a class is passed as an argument returns false if any of the collection is a member of such class

- [x] my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex

- [x] my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex

- [x] my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

- [x] my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+  # my_count
+  def my_count
+    index = 0
+    each do |_item|
+      index += 1
+    end
+    index
+  end
+  # my_count
```
Comment (mahmoud717) [inline main.rb]: - [x] my_count counts the number of items in enum that are equal to item if an argument is given

- [x] my_count counts the number of elements yielding a true value if a block is given

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      end
+      return new_arr
+    end
+
+    each do |item|
+      new_item = block.call(item)
+      new_arr << new_item
+    end
+    new_arr
+  end
```
Comment (mahmoud717) [inline main.rb]: - [x] my_map returns an Enumerator if no block is given

File: main.rb [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+  # my_inject
+  def my_inject(sign)
+    result = 0 if (sign == :+) || (sign == :-)
+    result = 1 if (sign == :*) || (sign == :/)
+    each do |item|
+      result = result.send(sign, item)
+    end
+    result
+  end
```
Comment (mahmoud717) [inline main.rb]: - [x] my_inject raises a "LocalJumpError" when no block or argument is given

- [x] my_inject searches for the longest word in an array of strings

- [x] my_inject does not mutate the original array

- [x] my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block

- [x] my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range

- [x] my_inject when a block is given with an initail value combines all elements of enum by applying a binary operation, specified by a block

- [x] my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method

- [x] my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method::range

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
### Changes required ♻️ 

**Hello @od-c0d3r & @wandji20**

Great job with the project so far, but there are still some changes to be made:
 
---
### Requested 🔴 

- [x] please check the inline comments
---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

**Comment by o-rashad** [comment]
Thank you @mahmoud717 for the review ⭐️
we are working on it and
considering pushing ASAP.

**Comment by o-rashad** [comment]
After the last review we added :

- For my_each, my_each_with_index, my_select, my_map
Returns an Enumerator if no block is given 

- For my_all?, my_any, my_none
Accepts blocks and arguments as a class, regex, or a pattern 

- For my_count
Counts the number of items that equal to the given argument.

**Review by umairarshadbutt — APPROVED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi @od-c0d3r  and @wandji20 
Good job with the changes! :tada: We are almost there.
We, however, do have a few changes to carry out before we proceed (see review comment).

##

 For this project, work on the suggested changes  👇.
- [x] Rename the file name from `main.rb` to `enumerables.rb`
## NOTE: These are the variables I used in the tests below: 👇 
```
ARRAY_SIZE = 100
LOWEST_VALUE = 0
HIGHEST_VALUE = 9

describe 'enumerables' do
  let(:array) { Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } }
  let(:block) { proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } }
  let(:words) { %w[dog door rod blade] }
  let(:range) { Range.new(5, 50) }
  let(:hash) { { a: 1, b: 2, c: 3, d: 4, e: 5 } }
  let(:numbers) { [1, 2i, 3.14] }
  let!(:array_clone) { array.clone }
```
## Failures: 10
- [x] #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
### Test 👉 `expect(array.my_all?(3)).to be array.all?(3)`
|Expected|Got|
|:---:|:---:|
|` false`|`true`|
- [x] #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
### Test 👉 `expect(true_array.my_any?).to be true_array.any?`
|Expected|Got|
|:---:|:---:|
|` true`|`false`|
- [x] #my_none? returns false if the block ever returns true for all elements
### Test 👉 `expect(array.my_none?(&false_block)).to eq(array.none?(&false_block))`
|Expected|Got|
|:---:|:---:|
|` false`|`true`|
- [x] #my_none? returns false if the block ever returns true for all elements::range
### Test 👉 `expect(range.my_none?(&false_block)).to eq(range.none?(&false_block))`
|Expected|Got|
|:---:|:---:|
|` false`|`true`|
- [x] #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
### Test 👉 `expect(words.my_none?(5)).to be words.none?(5)`
|Expected|Got|
|:---:|:---:|
|` false`|`true`|
- [x] #my_inject raises a "LocalJumpError" when no block or argument is given
### Test 👉 `expect { array.my_inject }.to raise_error(LocalJumpError)`
|Expected|Got|
|:---:|:---:|
|` LocalJumpError`|`Nothing was raised`|
- [x] #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range
### Test 👉 `actual = range.my_inject { |prod, n| prod * n }`
|Error Name|
|:---:|
|` undefined local variable or method `length' for 5..50:Range`|
- [x] #my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
### Test 👉 `actual = range.my_inject(4) { |prod, n| prod * n }`
|Error Name|
|:---:|
|` undefined local variable or method `length' for 5..50:Range`|
- [x] #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method:range
### Test 👉 `expect(range.my_inject(:*)).to eq range.inject(:*)`
|Expected|Got|
|:---:|:---:|
|` 1267253883404724085150525340252698701849068398706688000000000000`|`6336269417023620425752626701263493509245341993533440000000000000`|
- [x] #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
### Test 👉 `expect(actual).to eq 40`
|Expected|Got|
|:---:|:---:|
|` 40`|`80`|
##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

**Comment by o-rashad** [comment]
Thanks for the review @umairarshadbutt 

We are working on it and Pushing ASAP

**Review by PhillipUg — CHANGES_REQUESTED** [review]
Hey team, 👋

You have done a great job on this project so far! However, you still need to make some changes before we can approve this request 👇🏿 

### Changes Required ♻️ 

- [x] Some of your methods like `my_any`, `my_none`, and `my_inject`are still failing. So I am going to give you sample test cases below. Make sure they all pass.
```
# 5. my_any? (example test cases)
puts 'my_any?'
puts '-------'
p [7, 10, 3, 5].my_any?(&:even?) # => true
p [7, 10, 4, 5].my_any?(&:even?) # => true
p %w[q r s i].my_any? { |char| 'aeiou'.include?(char) } # => true
p [7, 11, 3, 5].my_any?(&:even?) # => false
p %w[q r s t].my_any? { |char| 'aeiou'.include?(char) } # => false
# test cases required by tse reviewer
p [3, 5, 4, 11].my_any? # => true
p [1, nil, false].my_any?(1) # => true
p [1, nil, false].my_any?(Integer) # => true
p %w[dog door rod blade].my_any?(/z/) # => false
p [1, 2, 3].my_any?(1) # => true
puts

# 6. my_none? (example test cases)
puts 'my_none?'
puts '--------'
p [3, 5, 7, 11].my_none?(&:even?) # => true
p %w[sushi pizza burrito].my_none? { |word| word[0] == 'a' } # => true
p [3, 5, 4, 7, 11].my_none?(&:even?) # => false
p %w[asparagus sushi pizza apple burrito].my_none? { |word| word[0] == 'a' } # => false
# test cases required by tse reviewer
p [1, 2, 3].my_none? # => false
p [1, 2, 3].my_none?(String) # => true
p [1, 2, 3, 4, 5].my_none?(2) # => false
p [1, 2, 3].my_none?(4) # => true
puts

# 9. my_inject
puts 'my_inject'
puts '---------'
p [1, 2, 3, 4].my_inject(10) { |accum, elem| accum + elem } # => 20
p [1, 2, 3, 4].my_inject { |accum, elem| accum + elem } # => 10
p [5, 1, 2].my_inject('+') # => 8
p (5..10).my_inject(2, :*) # should return 302400
p (5..10).my_inject(4) { |prod, n| prod * n } # should return 604800
```

#### Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack as **Phillip Musiime**
I would be happy to help! 👩‍💻 

Happy coding 🤓

**Comment by o-rashad** [comment]
Thank you @PhillipUg for the review ⭐️⭐️⭐️

Your provided test cases helped us identify the weak points within the code.

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Rashad Rashad & Wandji Wandji** 
Good job with the improvements to your project :muscle:  :muscle:
It meets most of the requirements :clap: :clap: :clap:  
There are just a few problems to correct before you can go ahead:

- [x] #my_inject raises a 'LocalJumpError' when no block or argument is given
    ```
    begin
      [1,2,3,4].my_inject
    rescue LocalJumpError
      true
    end
    ```

- [x] #my_inject when a block is given without an initail value combines all elements of enum
by applying a binary operation, specified by a block::range
      `(1..3).inject(&proc{|total, num| total*num}) == (1..3).my_inject(&proc{|total, num| total*num})`
- [x] #my_inject when a symbol is specified without an initail value combines each element of
the collection by applying the symbol as a named method
      `[1,2,3].inject(:+) == [1,2,3].my_inject(:+)`
- [x] #my_inject when a symbol is specified without an initail value combines each element of
the collection by applying the symbol as a named method:range
      `(1..9).inject(:+) == (1..9).my_inject(:+)`
- [x] #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
    `multiply_els([1,2,3]) == 6`
- [x] In the README file, you explain user has to run the main.rb file, but you don't have a file with that name in your project.
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

File: enumerables.rb [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+        result = first
+        range_arr = to_a
+        (1..size - 1).my_each { |item| result = yield(result, range_arr[item]) }
+      end
+    end
+    result
+  end
+
+  # end Module
+end
```
Comment (voscarmv) [inline enumerables.rb]: Your `my_inject` method returned the following errors:

```
Failures:

  1) enumerables #my_inject searches for the longest word in an array of strings
     Failure/Error: expect(words.my_inject(&search)).to eq words.inject(&search)
     
       expected: "blade"
            got: 5
     
       (compared using ==)
     # ./spec/enumerables_spec.rb:410:in `block (3 levels) in <top (required)>'

  2) enumerables #my_inject does not mutate the original array
     Failure/Error: array.my_inject { |num| num + 1 }
     
     NoMethodError:
       undefined method `init' for #<Array:0x00005637cd6c4d68>
     # ./enumerables.rb:155:in `my_inject'
     # ./spec/enumerables_spec.rb:414:in `block (3 levels) in <top (required)>'

  3) enumerables #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block
     Failure/Error: expect(array.my_inject(&operation)).to eq array.inject(&operation)
     
     NoMethodError:
       undefined method `init' for #<Array:0x00005637cd7bd670>
     # ./enumerables.rb:155:in `my_inject'
     # ./spec/enumerables_spec.rb:420:in `block (4 levels) in <top (required)>'

  4) enumerables #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range
     Failure/Error: actual = range.my_inject { |prod, n| prod * n }
     
     NoMethodError:
       undefined method `init' for 5..50:Range
     # ./enumerables.rb:155:in `my_inject'
     # ./spec/enumerables_spec.rb:424:in `block (4 levels) in <top (required)>'

  5) enumerables #my_inject when a block is given with an initail value combines all elements of enum by applying a binary operation, specified by a block
     Failure/Error: actual = range.my_inject(4) { |prod, n| prod * n }
     
     NoMethodError:
       undefined method `init' for 5..50:Range
     # ./enumerables.rb:155:in `my_inject'
     # ./spec/enumerables_spec.rb:432:in `block (4 levels) in <top (required)>'
```

Here's an explanation of each of these error messages:

1)  Your `my_inject` is returning the wrong output here. It should return the value of the longest word, not the number of letters in it. You may test your code in the following manner to make sure it outputs the correct value:

```
search = proc { |memo, word| memo.length > word.length ? memo : word }
words = %w[dog door rod blade] 
words.my_inject(&search) # This should return "blade"
```

2, 3, 4, 5: Apparently, you have a bug in line 155 in your `enumerables.rb` file. Please make sure that the `init` method you are using here exists for the cases. You may use the following setup for testing these cases:

```
array = [5, 4, 3, 2, 1]
range =   Range.new(5, 50)
operation = proc { |sum, n| sum + n }
array_clone = array.clone 

# Make sure the following calls to `my_inject` don't produce an `method init not found` error:
array.my_inject { |num| num + 1 }
array.my_inject(&operation)
range.my_inject { |prod, n| prod * n }
range.my_inject(4) { |prod, n| prod * n }
```

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Nice work so far, team. 👍 
## However, there are some changes required 👀 🧰 🔨 🔧 

- Please read my code comments in order to fix the errors found in your methods!

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Comment by o-rashad** [comment]
Thank you @voscarmv ⭐️⭐️⭐️
Well, descriptive and oriented review.
And yes for 2,3,4,5 was a missing indentation 😇

**Review by voscarmv — APPROVED** [review]
# Project status: Approved ✨ 😎👍 💯 ✔️
### You may merge this branch and move forward ⛙🛣
### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Pull request merged**